extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js

```diff
@@ -1161,3 +1318,51 @@ function getFirstInstanceValue_(instances, extractFn) {
   }
   return null;
 }
+
+/**
+ * Maps a sentinel value to an object consisting of the impl to which that
+ * sentinel value belongs and the corresponding message handler for that impl.
+ * @type{!Object<string, !Object>} */
+const fluidListeners = {};
+let listenerForFluidRegistered = false;
+
+/**
+ * @param {!AmpAdNetworkDoubleclickImpl} instance
+ */
+function registerListenerForFluid(instance) {
+  if (!fluidListeners[instance.sentinel]) {
+    fluidListeners[instance.sentinel] = {
+      instance,
+      connectionEstablished: false,
+    };
+  }
+  if (!listenerForFluidRegistered) {
+    instance.win.addEventListener('message', event => {
+      const data = tryParseJson(getData(event));
+      if (event.origin != SAFEFRAME_ORIGIN || !data || !data['sentinel']) {
+        return;
+      }
+      const listener = fluidListeners[data['sentinel']];
+      dev().assert(listener, 'postMessage listener does not exist');
+      if (data['s'] != 'creative_geometry_update') {
+        if (!listener.connectionEstablished) {
+          listener.instance.connectFluidMessagingChannel();
+          listener.connectionEstablished = true;
+        }
+        return;
+      }
+      listener.instance.receiveMessageForFluid(data);
+    }, false);
+    listenerForFluidRegistered = true;
+  }
+}
+
+/** @visibleForTesting */
+export function unregisterListenersForFluid() {
```

Comment:
Should this be called as part of unlayoutCallback?

Reply:
Maybe. This would unregister all listeners on the page, not just for this particular slot. Is that fine?

Reply:
Done.